### PR TITLE
Allow 'is Mixin' checks.

### PR DESCRIPTION
**Pull request by floitsch**
Also use flattened interfaces in tree grower.

### CODE REVIEW

File: src/compiler/mixin.cc [hunk 1]
```diff
@@ -77,7 +76,19 @@ static void add_stubs(ir::Class* klass) {
                                               forward_arguments,
                                               range);
 
-      auto stub = _new ir::MixinStub(method_name, klass, shape, method->range());
+      ir::MethodInstance* stub;
+      if (method->is_IsInterfaceOrMixinStub()) {
+        // We copy over the method (used to know if a class is an interface or mixin).
```
Comment (kasperl) [inline src/compiler/mixin.cc:81]:
```suggestion
        // We copy over the method (used to determine if a class is an interface or mixin).
```

File: src/compiler/tree.cc [hunk 2]
```diff
@@ -484,11 +484,25 @@ class Fixup : public ReplacingVisitor {
...
+    valid_check_targets_.insert_all(grown_classes);
 
     for (auto klass : grown_classes) {
       for (auto interface : klass->interfaces()) {
-        grown_classes_and_interfaces_.insert(interface);
+        valid_check_targets_.insert(interface);
+      }
+      for (auto current = klass; current != null; current = current->super()) {
+        if (current != klass && valid_check_targets_.contains(current)) {
+          // No need to duplicate work. The current class will (or was already) be
```
Comment (kasperl) [inline src/compiler/tree.cc:495]:
```suggestion
          // No need to duplicate work. The current class will be (or was already)
```

File: tests/mixin-is-test.toit [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+mixin MixC:
+  c-method: return 43
+
+mixin MixD extends MixB with MixC:
+  d-method: return 44
+
+mixin MixE:
+  e-method: return 45
+
```
Comment (kasperl) [inline tests/mixin-is-test.toit:21]: Drop one newline.

File: tests/mixin-is-test.toit [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+class ClassA extends Object with MixD MixE:
+
+class ClassB extends Object with MixD:
+
+confuse x -> any: return x
```
Comment (kasperl) [inline tests/mixin-is-test.toit:27]: You may want a better confuse function here. This is a bit too easy for the type propagator.

**Review by kasperl — APPROVED** [review]
Looks good to me.

File: src/compiler/mixin.cc (hunk 1 above)
Comment (floitsch) [reply src/compiler/mixin.cc:81]: done.

File: src/compiler/tree.cc (hunk 2 above)
Comment (floitsch) [reply src/compiler/tree.cc:495]: done.

File: tests/mixin-is-test.toit (hunk 3 above)
Comment (floitsch) [reply tests/mixin-is-test.toit:21]: done.

File: tests/mixin-is-test.toit (hunk 4 above)
Comment (floitsch) [reply tests/mixin-is-test.toit:27]: I think we should do that in a follow-up PR. For all `confuse` methods.

**Pull request merged**